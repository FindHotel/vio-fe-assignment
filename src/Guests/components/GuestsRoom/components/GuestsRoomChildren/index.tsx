import React, { ReactElement } from "react";
import { useSelector, useDispatch } from "react-redux";

import Button from "components/Button";
import Select, { Option } from "components/Select";
import { ReactComponent as IconPlus } from "assets/plus.svg";
import { ReactComponent as IconMinus } from "assets/minus.svg";
import {
  StoreStateType,
  RoomType,
  addChild,
  removeChild,
  removeSpecificChild,
  changeChildAge,
} from "store";

import {
  Guest,
  GuestText,
  GuestButtonGroup,
  KidsContainer,
  IconRemoveChild,
} from "../../styles";

type GuestsRoomChildrenProps = {
  position: number;
};

const GuestsRoomChildren = ({
  position,
}: GuestsRoomChildrenProps): ReactElement => {
  const dispatch = useDispatch();
  const { children }: RoomType = useSelector(
    (state: StoreStateType) => state.rooms[position]
  );

  return (
    <>
      <Guest>
        <GuestText>Children</GuestText>
        <GuestButtonGroup>
          <Button
            onClick={() => dispatch(removeChild(position))}
            theme="secondary"
            dataTestId="remove-child-button"
          >
            <IconMinus />
          </Button>
          <GuestText data-testid="children-counter">
            {children.length}
          </GuestText>
          <Button
            onClick={() => dispatch(addChild(position))}
            theme="secondary"
            dataTestId="add-child-button"
          >
            <IconPlus />
          </Button>
        </GuestButtonGroup>
      </Guest>
      <KidsContainer>
        {children.map((child, index) => (
          <Guest key={`child-${index}`}>
            <GuestText>Child {index + 1} age</GuestText>
            <GuestButtonGroup>
              <Select
                dataTestId={`child-${index}-select`}
                value={String(child.age)}
                onChange={(event) =>
                  dispatch(
                    changeChildAge({
                      childrenPosition: index,
                      roomPosition: position,
                      age: Number(event.target.value),
                    })
                  )
                }
              >
                <Option value="">Age</Option>
                {[
                  ...Array.from(Array(17).keys()).map((age) => (
                    <Option key={`child-age-${age}`} value={String(age)}>
                      {age}
                    </Option>
                  )),
                ]}
              </Select>
              <IconRemoveChild
                data-testid={`remove-child-${index}-icon`}
                onClick={() =>
                  dispatch(
                    removeSpecificChild({
                      childrenPosition: index,
                      roomPosition: position,
                    })
                  )
                }
              />
            </GuestButtonGroup>
          </Guest>
        ))}
      </KidsContainer>
    </>
  );
};

export default GuestsRoomChildren;
