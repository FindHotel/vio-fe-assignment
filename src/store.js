import * as R from 'ramda';
import { createStore, combineReducers } from 'redux';

import { deserializeRoomsData } from './libs/url';
import {
  MAX_ROOMS_COUNT, MAX_ADULTS_FOR_ROOM_COUNT, MAX_CHILDREN_FOR_ROOM_COUNT, MAX_ROOM_OCCUPANCY_COUNT,
} from './constants';

const initialRoomsState = [];

const roomPattern = {
  adults: 0,
  childrenAges: [],
};

// reducer

const roomsReducer = (state = initialRoomsState, action) => {
  const { type, payload } = action;
  const isRoomOccupancyExhausted = (roomId) => {
    const currentOccupancy = R.length(R.prop('childrenAges', state[roomId])) + R.prop('adults', state[roomId]);
    return currentOccupancy < MAX_ROOM_OCCUPANCY_COUNT;
  };

  switch (type) {
    case 'GET_ROOMS_FROM_URL': {
      const mapIndexed = R.addIndex(R.map);
      return mapIndexed((r, i) => R.assoc('id', `${i}`, r), deserializeRoomsData());
    }
    case 'ADD_ROOM':
      return state.length + 1 <= MAX_ROOMS_COUNT ? R.append(roomPattern, state) : state;
    case 'REMOVE_ROOM_BY_ID':
      return R.remove(payload, 1, state);
    case 'INC_ADULT_BY_ROOM_ID':
      if (
        state[payload].adults + 1 <= MAX_ADULTS_FOR_ROOM_COUNT
        && isRoomOccupancyExhausted(payload)
      ) {
        state[payload].adults += 1;
      }
      return R.update(payload, state[payload], state);
    case 'DEC_ADULT_BY_ROOM_ID':
      if (state[payload].adults - 1 >= 0) {
        state[payload].adults -= 1;
      }
      return R.update(payload, state[payload], state);
    case 'INC_CHILD_BY_ROOM_ID':
      if (
        R.length(R.prop('childrenAges', state[payload])) < MAX_CHILDREN_FOR_ROOM_COUNT
        && isRoomOccupancyExhausted(payload)
      ) {
        state[payload].childrenAges.push(null);
      }
      return R.update(payload, state[payload], state);
    case 'DEC_CHILD_BY_ROOM_ID':
      if (R.length(R.prop('childrenAges', state[payload])) > 0) {
        state[payload].childrenAges.pop();
      }
      return R.update(payload, state[payload], state);
    case 'REMOVE_CHILD_BY_ID': {
      const { roomId, childId } = payload;

      state[roomId].childrenAges = R.remove(childId, 1, R.prop('childrenAges', state[roomId]));
      return R.update(roomId, state[roomId], state);
    }
    case 'INC_CHILD_AGE_BY_ID': {
      const { roomId, childId } = payload;

      state[roomId].childrenAges[childId] += 1;
      return R.update(roomId, state[roomId], state);
    }
    case 'DEC_CHILD_AGE_BY_ID': {
      const { roomId, childId } = payload;

      if (state[roomId].childrenAges[childId] > 0) {
        state[roomId].childrenAges[childId] -= 1;
      }
      return R.update(roomId, state[roomId], state);
    }
    default:
      return state;
  }
};

// actions

export const getRoomsFromUrl = () => ({
  type: 'GET_ROOMS_FROM_URL',
});

export const addRoom = () => ({
  type: 'ADD_ROOM',
});

export const removeRoomById = (roomId) => ({
  type: 'REMOVE_ROOM_BY_ID',
  payload: roomId,
});

export const incAdultByRoomId = (roomId) => ({
  type: 'INC_ADULT_BY_ROOM_ID',
  payload: roomId,
});

export const decAdultByRoomId = (roomId) => ({
  type: 'DEC_ADULT_BY_ROOM_ID',
  payload: roomId,
});

export const incChildByRoomId = (roomId) => ({
  type: 'INC_CHILD_BY_ROOM_ID',
  payload: roomId,
});

export const decChildByRoomId = (roomId) => ({
  type: 'DEC_CHILD_BY_ROOM_ID',
  payload: roomId,
});

export const removeChildById = (roomId, childId) => ({
  type: 'REMOVE_CHILD_BY_ID',
  payload: { roomId, childId },
});

export const incChildAgeById = (roomId, childId) => ({
  type: 'INC_CHILD_AGE_BY_ID',
  payload: { roomId, childId },
});

export const decChildAgeById = (roomId, childId) => ({
  type: 'DEC_CHILD_AGE_BY_ID',
  payload: { roomId, childId },
});

export default createStore(
  combineReducers({ rooms: roomsReducer }),
  // eslint-disable-next-line no-underscore-dangle
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
);
