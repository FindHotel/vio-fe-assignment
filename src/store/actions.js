export const getRoomsFromUrl = () => ({
  type: 'GET_ROOMS_FROM_URL',
});

export const addRoom = () => ({
  type: 'ADD_ROOM',
});

export const removeRoomById = (roomId) => ({
  type: 'REMOVE_ROOM_BY_ID',
  payload: roomId,
});

export const incAdultByRoomId = (roomId) => ({
  type: 'INC_ADULT_BY_ROOM_ID',
  payload: roomId,
});

export const decAdultByRoomId = (roomId) => ({
  type: 'DEC_ADULT_BY_ROOM_ID',
  payload: roomId,
});

export const incChildByRoomId = (roomId) => ({
  type: 'INC_CHILD_BY_ROOM_ID',
  payload: roomId,
});

export const decChildByRoomId = (roomId) => ({
  type: 'DEC_CHILD_BY_ROOM_ID',
  payload: roomId,
});

export const removeChildById = (roomId, childId) => ({
  type: 'REMOVE_CHILD_BY_ID',
  payload: { roomId, childId },
});

export const incChildAgeById = (roomId, childId) => ({
  type: 'INC_CHILD_AGE_BY_ID',
  payload: { roomId, childId },
});

export const decChildAgeById = (roomId, childId) => ({
  type: 'DEC_CHILD_AGE_BY_ID',
  payload: { roomId, childId },
});

export const resetToInitial = () => ({
  type: 'GET_ROOMS_FROM_URL',
});
