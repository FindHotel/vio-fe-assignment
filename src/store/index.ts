import { createSlice, configureStore } from "@reduxjs/toolkit";

export type RoomType = {
  maxNumberOfAdults: number;
  maxNumberOfChildren: number;
  maxOccupancy: number;
  adultsCounter: number;
  totalGuests: number;
  children: [
    {
      age: number | null;
    }?
  ];
};

export type StoreStateType = {
  maxNumberOfRooms: number;
  rooms: RoomType[];
};

const initialRoomsState = {
  maxNumberOfAdults: 5,
  maxNumberOfChildren: 3,
  maxOccupancy: 5,
  adultsCounter: 1,
  totalGuests: 1,
  children: [],
};

const initialState = {
  maxNumberOfRooms: 8,
  rooms: [initialRoomsState],
};

const counterSlice = createSlice({
  name: "rooms",
  initialState,
  reducers: {
    resetStore: (state): void => {
      state.rooms = [initialRoomsState];
    },
    addRoom: (state): void => {
      const { rooms, maxNumberOfRooms } = state;

      if (rooms.length < maxNumberOfRooms) {
        rooms.push(initialRoomsState);
      }
    },
    removeRoom: (state, action): void => {
      if (state.rooms.length > 1) {
        state.rooms.splice(action.payload, 1);
      }
    },
    addAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.adultsCounter < room.maxNumberOfAdults &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.adultsCounter += 1;
        room.totalGuests += 1;
      }
    },
    removeAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (room.adultsCounter > 1) {
        room.adultsCounter -= 1;
        room.totalGuests -= 1;
      }
    },
    addChild: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.children.length < room.maxNumberOfChildren &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.children.push({
          age: null,
        });
        room.totalGuests += 1;
      }
    },
    removeChild: (state, action): void => {
      const room = state.rooms[action.payload];
      const { children } = room;

      if (children.length) {
        children.splice(children.length - 1, 1);
        room.totalGuests -= 1;
      }
    },
    removeSpecificChild: (state, action): void => {
      const { childrenPosition, roomPosition } = action.payload;
      const room = state.rooms[roomPosition];
      const { children } = room;

      if (children.length) {
        children?.splice(childrenPosition, 1);
        room.totalGuests -= 1;
      }
    },
    changeChildAge: (state, action): void => {
      const { childrenPosition, roomPosition, age } = action.payload;
      const { children } = state.rooms[roomPosition];

      if (Boolean(age)) {
        children[childrenPosition].age = Number(age);
      }
    },
  },
});

export const {
  resetStore,
  addRoom,
  removeRoom,
  addAdult,
  removeAdult,
  addChild,
  removeChild,
  removeSpecificChild,
  changeChildAge,
} = counterSlice.actions;

const store = configureStore({
  reducer: counterSlice.reducer,
});

export default store;
