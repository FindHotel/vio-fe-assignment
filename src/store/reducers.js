import * as R from 'ramda';

import {
  MAX_CHILDREN_FOR_ROOM_COUNT,
  MAX_ROOM_OCCUPANCY_COUNT,
  MAX_ROOMS_COUNT,
  MIN_ADULTS_FOR_ROOM_COUNT,
} from '../constants';
import { getRoomsDataFromQueryParams } from '../libs/url';

const initialRoomsState = [];

const roomPattern = {
  adults: MIN_ADULTS_FOR_ROOM_COUNT,
  childrenAges: [],
};

const validateRoomsData = (rooms) => {
  const validateRoom = (room) => {
    const childrenCount = R.length(R.prop('childrenAges', room) || []);
    const adultsCount = R.prop('adults', room);
    const occupancy = childrenCount + adultsCount;

    if (
      occupancy <= MAX_ROOM_OCCUPANCY_COUNT
      && childrenCount <= MAX_CHILDREN_FOR_ROOM_COUNT
      && childrenCount >= 0
      && adultsCount >= MIN_ADULTS_FOR_ROOM_COUNT
    ) {
      return true;
    }

    console.log('Room data is not valide', room);
    return false;
  };

  return R.filter(validateRoom, rooms);
};

// Selectors

const isIncOccupancyDisabled = (rooms, roomId) => {
  const currentOccupancy = R.length(R.prop('childrenAges', rooms[roomId]) || []) + R.prop('adults', rooms[roomId]);
  return currentOccupancy >= MAX_ROOM_OCCUPANCY_COUNT;
};

export const isAdultsIncDisabled = (state, roomId) => {
  const rooms = R.prop('rooms', state);
  const room = rooms[roomId];
  return isIncOccupancyDisabled(rooms, roomId) || R.prop('adults', room) >= MAX_ROOM_OCCUPANCY_COUNT;
};

export const isAdultsDecDisabled = (state, roomId) => {
  const rooms = R.prop('rooms', state);
  const room = rooms[roomId];
  return R.prop('adults', room) <= MIN_ADULTS_FOR_ROOM_COUNT;
};

export const isChildrenIncDisabled = (state, roomId) => {
  const rooms = R.prop('rooms', state);
  const room = rooms[roomId];
  return isIncOccupancyDisabled(rooms, roomId) || R.length(R.prop('childrenAges', room)) >= MAX_CHILDREN_FOR_ROOM_COUNT;
};

export const isRoomRemovingDisabled = (roomId) => Number(roomId) === 0;

// Reducer

export const roomsReducer = (state = initialRoomsState, action) => {
  const { type, payload } = action;

  switch (type) {
    case 'GET_ROOMS_FROM_URL': {
      const roomsData = getRoomsDataFromQueryParams();
      const validatedRoomsData = validateRoomsData(roomsData);
      return R.length(validatedRoomsData) ? validatedRoomsData : [R.clone(roomPattern)];
    }

    case 'ADD_ROOM':
      return state.length + 1 <= MAX_ROOMS_COUNT ? R.append(R.clone(roomPattern), state) : state;

    case 'REMOVE_ROOM_BY_ID':
      return R.remove(payload, 1, state);

    case 'INC_ADULT_BY_ROOM_ID':
      if (!isIncOccupancyDisabled(state, payload)) {
        state[payload].adults += 1;
      }
      return R.update(payload, state[payload], state);

    case 'DEC_ADULT_BY_ROOM_ID':
      if (state[payload].adults - 1 >= MIN_ADULTS_FOR_ROOM_COUNT) {
        state[payload].adults -= 1;
      }
      return R.update(payload, state[payload], state);

    case 'INC_CHILD_BY_ROOM_ID':
      if (
        R.length(R.prop('childrenAges', state[payload])) < MAX_CHILDREN_FOR_ROOM_COUNT
        && !isIncOccupancyDisabled(state, payload)
      ) {
        state[payload].childrenAges.push(null);
      }
      return R.update(payload, state[payload], state);

    case 'DEC_CHILD_BY_ROOM_ID':
      if (R.length(R.prop('childrenAges', state[payload])) > 0) {
        state[payload].childrenAges.pop();
      }
      return R.update(payload, state[payload], state);

    case 'REMOVE_CHILD_BY_ID': {
      const { roomId, childId } = payload;
      state[roomId].childrenAges = R.remove(childId, 1, R.prop('childrenAges', state[roomId]));
      return R.update(roomId, state[roomId], state);
    }

    case 'INC_CHILD_AGE_BY_ID': {
      const { roomId, childId } = payload;
      state[roomId].childrenAges[childId] += 1;
      return R.update(roomId, state[roomId], state);
    }

    case 'DEC_CHILD_AGE_BY_ID': {
      const { roomId, childId } = payload;
      if (state[roomId].childrenAges[childId] > 0) {
        state[roomId].childrenAges[childId] -= 1;
      }
      return R.update(roomId, state[roomId], state);
    }

    default:
      return state;
  }
};
