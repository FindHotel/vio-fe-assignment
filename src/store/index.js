import { combineReducers, createStore } from 'redux';

import { roomsReducer } from './reducers';

export default createStore(
  combineReducers({
    rooms: roomsReducer,
  }),
  // eslint-disable-next-line no-underscore-dangle
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
);
