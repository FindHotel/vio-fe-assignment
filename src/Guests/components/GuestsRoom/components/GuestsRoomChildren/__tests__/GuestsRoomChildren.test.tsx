import React from "react";
import { Provider } from "react-redux";
import { render, fireEvent, screen } from "@testing-library/react";

import store, { resetStore } from "store";

import GuestsRoomChildren from "..";

describe("<GuestsRoomChildren />", () => {
  beforeEach(() => store.dispatch(resetStore()));

  it("should render properly", () => {
    const { container } = render(
      <Provider store={store}>
        <GuestsRoomChildren position={0} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should be able to add and remove a child", () => {
    render(
      <Provider store={store}>
        <GuestsRoomChildren position={0} />
      </Provider>
    );

    const childrenCounter = screen.getByTestId("children-counter");

    expect(childrenCounter.textContent).toBe("0");

    fireEvent.click(screen.getByTestId("add-child-0-button"));
    expect(childrenCounter.textContent).toBe("1");

    fireEvent.click(screen.getByTestId("remove-child-0-button"));
    expect(childrenCounter.textContent).toBe("0");
  });

  it("should be able to change a child age", () => {
    const childAge = 3;

    render(
      <Provider store={store}>
        <GuestsRoomChildren position={0} />
      </Provider>
    );

    fireEvent.click(screen.getByTestId("add-child-0-button"));
    fireEvent.change(screen.getByTestId("child-0-0-select"), {
      target: { value: childAge },
    });

    const child = store.getState().rooms[0].children[0];

    expect(child.age).toBe(childAge);
  });
});
