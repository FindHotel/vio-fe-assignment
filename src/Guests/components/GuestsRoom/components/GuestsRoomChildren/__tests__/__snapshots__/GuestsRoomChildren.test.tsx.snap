// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GuestsRoomChildren /> should render properly 1`] = `
<div
  class="sc-gKckTs dPQnl"
>
  <span
    class="sc-iCfLBT JjZsB"
  >
    Children
  </span>
  <div
    class="sc-furvIG cWguzl"
  >
    <button
      class="sc-bdvvaa eWedmN"
      data-testid="remove-child-0-button"
    >
      <svg>
        minus.svg
      </svg>
    </button>
    <span
      class="sc-iCfLBT JjZsB"
      data-testid="children-counter"
    >
      0
    </span>
    <button
      class="sc-bdvvaa eWedmN"
      data-testid="add-child-0-button"
    >
      <svg>
        plus.svg
      </svg>
    </button>
  </div>
</div>
`;
